### PR TITLE
feat(aws-apigateway): add support for UsagePlan, ApiKey, UsagePlanKey

### PR DESCRIPTION
**Pull request by orangewise**
----

#723 

### Pull Request Checklist

* [ ] Testing
  - Unit test added (prefer not to modify an existing test, otherwise, it's probably a breaking change)
  - __CLI change?:__ coordinate update of integration tests with team
  - __cdk-init template change?:__ coordinated update of integration tests with team
* [ ] Docs
  - __jsdocs__: All public APIs documented
  - __README__: README and/or documentation topic updated
  - __Design__: For significant features, design document added to `design` folder
* [ ] Title and Description
  - __Change type__: title prefixed with **fix**, **feat** and module name in parens, which will appear in changelog
  - __Title__: use lower-case and doesn't end with a period
  - __Breaking?__: last paragraph: "BREAKING CHANGE: <describe what changed + link for details>"
  - __Issues__: Indicate issues fixed via: "**Fixes #xxx**" or "**Closes #xxx**"
* [ ] Sensitive Modules (requires 2 PR approvers)
  - IAM Policy Document (in @aws-cdk/aws-iam)
  - EC2 Security Groups and ACLs (in @aws-cdk/aws-ec2)
  - Grant APIs (only if not based on official documentation with a reference)

----

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license.

### CODE REVIEW

File: packages/@aws-cdk/aws-apigateway/README.md [hunk 1]
```diff
@@ -104,6 +104,57 @@ book.addMethod('GET', getBookIntegration, {
...
+  handler: 'hello.handler',
+  code: lambda.Code.asset('lambda')
+});
+
+const api = new apigateway.RestApi(this, 'hello-api', { });
+const integration = new apigateway.LambdaIntegration(hello);
+
+const v1 = api.root.addResource('v1');
+const echo = v1.addResource('echo');
+const echoMethod: apigateway.Method = echo.addMethod('GET', integration, { apiKeyRequired: true });
```
Comment (eladb) [inline packages/@aws-cdk/aws-apigateway/README.md]: the explicit type is not needed

File: packages/@aws-cdk/aws-apigateway/README.md [hunk 2]
```diff
@@ -104,6 +104,57 @@ book.addMethod('GET', getBookIntegration, {
...
+
+const usagePlan = new apigateway.UsagePlan(this, 'UsagePlan', {
```
Comment (eladb) [inline packages/@aws-cdk/aws-apigateway/README.md]: Does it make sense to offer a convinience method "restApi.addUsagePlan(id, options)`.

File: packages/@aws-cdk/aws-apigateway/README.md [hunk 3]
```diff
@@ -104,6 +104,57 @@ book.addMethod('GET', getBookIntegration, {
...
+      ]
+    }
+  ]
+});
+
+const key = new apigateway.ApiKey(this, 'ApiKey', {
+  resources: [api]
+});
+
+usagePlan.addApiKey(key);
```
Comment (eladb) [inline packages/@aws-cdk/aws-apigateway/README.md]: It should also be possible to specify `keys` in `UsagePlanProps`

File: packages/@aws-cdk/aws-apigateway/README.md [hunk 4]
```diff
@@ -104,6 +104,57 @@ book.addMethod('GET', getBookIntegration, {
...
+            rateLimit: 10,
+            burstLimit: 2
+          }
+        }
+      ]
+    }
+  ]
+});
+
+const key = new apigateway.ApiKey(this, 'ApiKey', {
```
Comment (eladb) [inline packages/@aws-cdk/aws-apigateway/README.md]: Add convenience `restApi.addApiKey(id);`

File: packages/@aws-cdk/aws-apigateway/README.md [hunk 5]
```diff
@@ -104,6 +104,57 @@ book.addMethod('GET', getBookIntegration, {
...
+  description: 'Free tier monthly usage plan',
+  quota: {
+    limit: 10000,
+    period: apigateway.Period.Month
+  },
+  throttle: {
+    rateLimit: 50,
+    burstLimit: 5
+  },
+  apiStages: [
```
Comment (eladb) [inline packages/@aws-cdk/aws-apigateway/README.md]: Should also offer `addStage`

File: packages/@aws-cdk/aws-apigateway/README.md [hunk 6]
```diff
@@ -104,6 +104,57 @@ book.addMethod('GET', getBookIntegration, {
...
+    {
+      api: api,
```
Comment (eladb) [inline packages/@aws-cdk/aws-apigateway/README.md]: This is not needed, you can deduce the `api` from the stage object

File: packages/@aws-cdk/aws-apigateway/README.md [hunk 7]
```diff
@@ -104,6 +104,57 @@ book.addMethod('GET', getBookIntegration, {
...
+      stage: api.deploymentStage,
+      throttle: [
```
Comment (eladb) [inline packages/@aws-cdk/aws-apigateway/README.md]: It's weird that we have `throttle` inside `throttle`

Comment (eladb) [reply packages/@aws-cdk/aws-apigateway/README.md]: Maybe it should be `new apigw.MethodThrottle(method, { rate, burst })`?

File: packages/@aws-cdk/aws-apigateway/README.md (hunk 2 above)
Comment (orangewise) [reply packages/@aws-cdk/aws-apigateway/README.md]: You're right, makes sense.

File: packages/@aws-cdk/aws-apigateway/README.md (hunk 4 above)
Comment (orangewise) [reply packages/@aws-cdk/aws-apigateway/README.md]: done.

File: packages/@aws-cdk/aws-apigateway/README.md (hunk 5 above)
Comment (orangewise) [reply packages/@aws-cdk/aws-apigateway/README.md]: I need some help on this one, if possible ;)

File: packages/@aws-cdk/aws-apigateway/README.md (hunk 6 above)
Comment (orangewise) [reply packages/@aws-cdk/aws-apigateway/README.md]: done.

File: packages/@aws-cdk/aws-apigateway/README.md (hunk 7 above)
Comment (orangewise) [reply packages/@aws-cdk/aws-apigateway/README.md]: I've removed one in the readme...

**Comment by orangewise** [comment]
Hi @eladb, thanks for reviewing. I've made a couple of changes you suggested. Could not wrap my head around the `addStage` stuff. Maybe you can help me out a little?

File: packages/@aws-cdk/aws-apigateway/lib/api-key.ts [hunk 8]
```diff
@@ -0,0 +1,117 @@
+import { CfnOutput, Construct, IResource as IResourceBase, Resource } from '@aws-cdk/cdk';
+import { CfnApiKey } from './apigateway.generated';
+import { ResourceOptions } from "./resource";
+import { RestApi } from './restapi';
+
+export interface ApiKeyAttributes {
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-apigateway/lib/api-key.ts]: Docstrings on public type.

File: packages/@aws-cdk/aws-apigateway/lib/api-key.ts [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+  /**
+   * The API key ID.
+   * @attribute
+   */
+  readonly keyId: string;
+}
+
+export interface IApiKey extends IResourceBase {
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-apigateway/lib/api-key.ts]: Docstrings on public type.

File: packages/@aws-cdk/aws-apigateway/lib/api-key.ts [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+  readonly keyId: string;
+
+  /**
+   * Exports a API key resource from this stack.
+   * @returns API key props that can be imported to another stack.
+   */
+  export(): ApiKeyAttributes;
+}
+
+export interface ApiKeyProps extends ResourceOptions {
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-apigateway/lib/api-key.ts]: Docstrings on public type.

File: packages/@aws-cdk/aws-apigateway/lib/api-key.ts [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+    });
+
+    this.keyId = resource.ref;
+  }
+
+  /**
+   * Exports a API key resource from this stack.
+   * @returns API key props that can be imported to another stack.
+   */
+  public export(): ApiKeyAttributes {
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-apigateway/lib/api-key.ts]: `export()`s are no longer required (or allowed, even). Please get rid of it. If you get compilation errors, please merge from `master`.

File: packages/@aws-cdk/aws-apigateway/lib/restapi.ts [hunk 12]
```diff
@@ -267,6 +269,26 @@ export class RestApi extends Resource implements IRestApi {
     return this.deploymentStage.urlForPath(path);
   }
 
+  /**
+   * Adds a usage plan.
+   *
+   * @param id
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-apigateway/lib/restapi.ts]: Either add docs or remove tags.

File: packages/@aws-cdk/aws-apigateway/lib/restapi.ts [hunk 13]
```diff
@@ -267,6 +269,26 @@ export class RestApi extends Resource implements IRestApi {
...
+   * @param props
+   */
+  public addUsagePlan(id: string, props: UsagePlanProps): UsagePlan {
+    return new UsagePlan(this, id, props);
+  }
+
+  /**
+   * Add an ApiKey
+   * @param id
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-apigateway/lib/restapi.ts]: Either add docs or remove tags.

File: packages/@aws-cdk/aws-apigateway/lib/usage-plan.ts [hunk 14]
```diff
@@ -0,0 +1,241 @@
...
+
+/**
+ * Container for defining throttling parameters to API stages or methods.
+ * @link https://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-request-throttling.html
+ */
+export interface ThrottleSettings {
+  /**
+   * The API request steady-state rate limit (average requests per second over an extended period of time)
+   *
+   * Type: Integer
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-apigateway/lib/usage-plan.ts]: Remove type, add `@default` tag.

File: packages/@aws-cdk/aws-apigateway/lib/usage-plan.ts [hunk 15]
```diff
@@ -0,0 +1,241 @@
...
+   */
+  readonly rateLimit?: number;
+
+  /**
+   * The maximum API request rate limit over a time ranging from one to a few seconds.
+   *
+   * Type: Integer
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-apigateway/lib/usage-plan.ts]: Remove type, add `@default` tag.

File: packages/@aws-cdk/aws-apigateway/lib/usage-plan.ts [hunk 16]
```diff
@@ -0,0 +1,241 @@
...
+}
+
+/**
+ * Specifies the maximum number of requests that clients can make to API Gateway APIs.
+ */
+export interface QuotaSettings {
+  /**
+   * The maximum number of requests that users can make within the specified time period.
+   *
+   * Type: Integer
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-apigateway/lib/usage-plan.ts]: Remove type, add `@default` tag.

File: packages/@aws-cdk/aws-apigateway/lib/usage-plan.ts [hunk 17]
```diff
@@ -0,0 +1,241 @@
...
+   */
+  readonly limit?: number;
+
+  /**
+   * For the initial time period, the number of requests to subtract from the specified limit.
+   *
+   * Type: Integer
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-apigateway/lib/usage-plan.ts]: Remove type, add `@default` tag.

File: packages/@aws-cdk/aws-apigateway/lib/usage-plan.ts [hunk 18]
```diff
@@ -0,0 +1,241 @@
...
+   */
+  readonly offset?: number;
+
+  /**
+   * The time period for which the maximum limit of requests applies.
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-apigateway/lib/usage-plan.ts]: Add `@default`

File: packages/@aws-cdk/aws-apigateway/lib/usage-plan.ts [hunk 19]
```diff
@@ -0,0 +1,241 @@
...
+/**
+ * Represents per-method throttling for a resource.
+ */
+export interface ThrottlingPerMethod {
+  /**
+   *
+   * [disable-awslint:ref-via-interface]
+   */
+  readonly method: Method,
+  readonly throttle: ThrottleSettings
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-apigateway/lib/usage-plan.ts]: Public docs please.

File: packages/@aws-cdk/aws-apigateway/lib/usage-plan.ts [hunk 20]
```diff
@@ -0,0 +1,241 @@
...
+export interface UsagePlanPerApiStage {
+
+  readonly api?: IRestApi,
+
+  /**
+   *
+   * [disable-awslint:ref-via-interface]
+   */
+  readonly stage?: Stage,
+  readonly throttle?: ThrottlingPerMethod[]
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-apigateway/lib/usage-plan.ts]: Public docs please.

File: packages/@aws-cdk/aws-apigateway/lib/usage-plan.ts [hunk 21]
```diff
@@ -0,0 +1,241 @@
...
+export enum UsagePlanKeyType {
+  ApiKey = 'API_KEY'
+}
+
+/**
+ * Represents the API stages that a usage plan applies to.
+ */
+export interface UsagePlanPerApiStage {
+
+  readonly api?: IRestApi,
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-apigateway/lib/usage-plan.ts]: Public docs please.

Comment (rix0rrr) [reply packages/@aws-cdk/aws-apigateway/lib/usage-plan.ts]: `@default`

File: packages/@aws-cdk/aws-apigateway/lib/usage-plan.ts [hunk 22]
```diff
@@ -0,0 +1,241 @@
...
+
+  /**
+   *
+   * [disable-awslint:ref-via-interface]
+   */
+  readonly stage?: Stage,
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-apigateway/lib/usage-plan.ts]: `@default`

File: packages/@aws-cdk/aws-apigateway/lib/usage-plan.ts (hunk 20 above)
Comment (rix0rrr) [reply packages/@aws-cdk/aws-apigateway/lib/usage-plan.ts]: `@default`

File: packages/@aws-cdk/aws-apigateway/lib/usage-plan.ts [hunk 23]
```diff
@@ -0,0 +1,241 @@
...
+  readonly throttle?: ThrottlingPerMethod[]
+}
+
+export interface UsagePlanProps {
+  /**
+   * API Stages to be associated which the usage plan.
+   */
+  readonly apiStages?: UsagePlanPerApiStage[],
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-apigateway/lib/usage-plan.ts]: `@default`

File: packages/@aws-cdk/aws-apigateway/lib/usage-plan.ts [hunk 24]
```diff
@@ -0,0 +1,241 @@
...
+
+  /**
+   * Represents usage plan purpose.
+   */
+  readonly description?: string,
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-apigateway/lib/usage-plan.ts]: `@default`

File: packages/@aws-cdk/aws-apigateway/lib/usage-plan.ts [hunk 25]
```diff
@@ -0,0 +1,241 @@
...
+
+  /**
+   * Number of requests clients can make in a given time period.
+   */
+  readonly quota?: QuotaSettings
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-apigateway/lib/usage-plan.ts]: `@default`

File: packages/@aws-cdk/aws-apigateway/lib/usage-plan.ts [hunk 26]
```diff
@@ -0,0 +1,241 @@
...
+
+  /**
+   * Overall throttle settings for the API.
+   */
+  readonly throttle?: ThrottleSettings,
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-apigateway/lib/usage-plan.ts]: `@default`

File: packages/@aws-cdk/aws-apigateway/lib/usage-plan.ts [hunk 27]
```diff
@@ -0,0 +1,241 @@
...
+
+  /**
+   * Name for this usage plan.
+   */
+  readonly name?: string
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-apigateway/lib/usage-plan.ts]: `@default`

File: packages/@aws-cdk/aws-apigateway/lib/usage-plan.ts [hunk 28]
```diff
@@ -0,0 +1,241 @@
...
+
+  /**
+   * ApiKey to be associated with the usage plan.
+   */
+  readonly apiKey?: IApiKey
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-apigateway/lib/usage-plan.ts]: `@default`

File: packages/@aws-cdk/aws-apigateway/lib/usage-plan.ts [hunk 29]
```diff
@@ -0,0 +1,241 @@
...
+}
+
+export class UsagePlan extends Resource {
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-apigateway/lib/usage-plan.ts]: Public docstring

File: packages/@aws-cdk/aws-apigateway/lib/usage-plan.ts [hunk 30]
```diff
@@ -0,0 +1,241 @@
...
+  /**
+   * @attribute
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-apigateway/lib/usage-plan.ts]: Public docstring

File: packages/@aws-cdk/aws-apigateway/lib/usage-plan.ts [hunk 31]
```diff
@@ -0,0 +1,241 @@
...
+    this.usagePlanId = resource.ref;
+
+    // Add ApiKey when
+    if (props.apiKey) {
+      this.addApiKey(props.apiKey);
+    }
+  }
+
+  /**
+   * Adds and ApiKey.
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-apigateway/lib/usage-plan.ts]: Typo

File: packages/aws-cdk/aws-cdk [hunk 32]
```diff
@@ -0,0 +1 @@
+/Users/ldrhgp1/ronald/cdk/aws-cdk/node_modules/aws-cdk
```
Comment (rix0rrr) [inline packages/aws-cdk/aws-cdk]: Oops?

File: packages/@aws-cdk/aws-apigateway/lib/api-key.ts [hunk 33]
```diff
@@ -0,0 +1,117 @@
...
+   * The API key ID.
+   * @attribute
+   */
+  readonly keyId: string;
+
+  /**
+   * Exports a API key resource from this stack.
+   * @returns API key props that can be imported to another stack.
+   */
+  export(): ApiKeyAttributes;
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-apigateway/lib/api-key.ts]: Can get rid of export()

File: packages/@aws-cdk/aws-apigateway/lib/usage-plan.ts [hunk 34]
```diff
@@ -0,0 +1,241 @@
...
+  /**
+   * @attribute
+   */
+  public readonly usagePlanId: string;
+
+  constructor(scope: Construct, id: string, props: UsagePlanProps = { }) {
+    super(scope, id);
+    let resource: CfnUsagePlan;
+
+    if (props !== undefined) {
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-apigateway/lib/usage-plan.ts]: Doesn't look like `props` can ever be undefined, so this `if` seems unnecessary.

File: packages/@aws-cdk/aws-apigateway/lib/usage-plan.ts [hunk 35]
```diff
@@ -0,0 +1,241 @@
...
+      resource = new CfnUsagePlan(this, 'Resource', {
+        apiStages: this.renderApiStages(props.apiStages),
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-apigateway/lib/usage-plan.ts]: For the `addApiStage` method, do this:

```ts
class UsagePlan {
  private readonly apiStages = new Array<UsagePlanPerApiStage>();

  constructor(...) {
    // ...
    resource = new CfnUsagePlan(this, 'Resource', {
      apiStages: new cdk.Token(() => this.renderApiStages(this.apiStages),
    });
 
    this.apiStages.push(...(props.apiStages || []));

    // ...
  }

  public addApiStage(apiStage: UsagePlanPerApiStage) {
    this.apiStages.push(apiStage);
  }

  // ...
}
```

Comment (orangewise) [reply packages/@aws-cdk/aws-apigateway/lib/usage-plan.ts]: Awesome, this works :)

**Comment by orangewise** [comment]
Hi @rix0rrr, I've updated the PR, maybe you can check when you have time.

Thanks!

**Comment by rix0rrr** [comment]
Sorry, this has slipped my attention. Could you merge from master to make sure this still passes linter checks? We've added a couple in the mean time and I'd hate for a merge to break the build.

**Comment by orangewise** [comment]
Np, I've pulled in the changes...

**Pull request merged**